mlton/backend/ssa2-to-rssa.fun

```diff
@@ -1642,38 +1643,255 @@ fun convert (program as S.Program.T {functions, globals, main, ...},
                           returns = transTypes returns,
                           start = start}
          end
+      fun translateGlobalStatics statements
+         : (Var.t * Type.t * Var.t Static.t) vector * S.Statement.t vector
+         =
+         let
+            val statics = ref []
+            val keeps = ref []
+            datatype elem = datatype Static.Data.elem
+            datatype z = datatype PackedRepresentation.staticOrElem
+            val {get = globalStatic: Var.t -> Var.t elem option,
+                 set = setGlobalStatic, destroy = destGlobalStatics} =
+               Property.destGetSetOnce
+               (Var.plist, Property.initConst NONE)
+            fun pushStatic (v, ty, rty, s) =
+               (setVarInfo (v, {ty=ty});
+                List.push (statics, (v, rty, s)))
+            fun pushKeep st = List.push (keeps, st)
+
+            fun validArg {elt, isMutable} =
+               case !Control.staticAllocInternalPtrs of
```

Comment:
Are the `staticAllocInternalPtrs` variants misnamed?  Should `Control.Static` instead by `Control.Immutable`?  That seems to be more in line with the condition.  Also, the distinction between `Control.All` and `Control.Static` is unclear, because an object with non-static ptr fields cannot be statically allocated.

Reply:
No. This setting is controlling where internal pointers on a static object may point. Since the current strategy is to skip over statics, we allow pointers only to other statics, else an object may fail to be collected. It is relevant only for mutable pointer fields, since initialization must use statics regardless (since it must be available at link-time).

So this is what condition we place on pointers internal to statically-allocated objects
If it is All: then we may have down-pointers from the statics to the heap
If it is Static: we may have pointers from statics into statics
If it is None: we may not have pointers from statics  (this is mostly for completeness)

Reply:
Arguably if it were All, we could allow invalid pointers by adding extra initialization steps after the Heap is set up, but this is an optimization for a case we don't currently support.

Reply:
I see; `All` is really unsafe with the current implementation.